### ISSUE TITLE
[Feature Request] Tidal Support

### ISSUE THREAD

**Issue by warllo54** [jaedb/Iris#363]
Is there any way you could add tidal support. There is a mopidy-tidal component that works great with other front ends however other front ends lack so many other features.

**Comment by jaedb** [comment]
Iris supports most backends out-of-the-box. You can find these under _My Music > Browse_, so yes, Iris should already support Tidal.

Is the [_Mopidy-Tidal_](https://github.com/mones88/mopidy-tidal) extension you refer to the one that has been archived?

**Comment by warllo54** [comment]
Bummer. I wasn't aware that it was archived. It has been working pretty well for me. Hopefully Tidal doesn't change anything that will break it or hopefully someone will fork it and maintain it. I would but I don't have the skills to do it. Thanks for the quick response.

**Comment by warllo54** [comment]
I did get a chance to test Tidal and the basics work great but some artwork doesn't display and there isn't a good way to add music to your library but searching, playing, and adding to playlist work great. Thanks for an amazing project. I made a small donation it's not much but thank you again for a great project.

**Comment by jaedb** [comment]
@warllo54 thanks heaps for your support - every little bit counts!

The biggest hurdle with Tidal is there complete lack of documentation of their API. There are a few API wrappers floating around in GitHub which look great, but with not much activity. This makes me wary of committing effort to integrating Iris to a poorly-supported API.

I'll keep an eye out for ways that this might work though.

**Comment by warllo54** [comment]
@jaedb I recently discovered Qobuz which offers many of the same features as Tidal, but they have an open api and encourage 3rd party apps. Might be a better option than Tidal. Here is a link to the public api https://github.com/Qobuz/api-documentation

I haven't been able to find a python module for it but the api is very well documented.

**Comment by g4v4g4i** [comment]
There is: https://github.com/secondwtq/mopidy-qobuz

**Comment by jaedb** [comment]
@warllo54 that API Documentation repo doesn't appear to exist any more under that account (I can only find one under another person's GitHub: https://github.com/jonaski/qobuz-api-documentation). I wonder if this suggests they aren't making their API as accessible as it once was.

In the meantime, _My Music > Browse_ is where you can find the content for any of your installed Mopidy extensions, so you will still be able to use Qobuz, Tidal, etc in that way.

Right now there doesn't appear to be much demand for a full API integration (like with Spotify) to Iris, so I can't justify prioritizing it at this time.

**Comment by jaedb** [comment]
Closing as this is not viable with the current tools available.

**Comment by sterni1971** [comment]
I (like many other) decided today to quit my spotify sub. The tidal web ui is really nice, but I'm missing IRIS support (I know the basic support is under browse)
The mopidy-tidal and the unofficial python api also looks quite active. So I would vote for re-open the feature request :-)

**Comment by kingosticks** [comment]
While Spotify has issues (both political and technical), Tidal doesn't even have a public API. Nor have they shown any interest in providing one. So you get all the issues a reverse-engineered solution suffers from i.e. it breaks when Tidal randomly change their security/implementation. There's a recent example [here](https://github.com/tehkillerbee/mopidy-tidal/issues/43). From a purely technical standpoint, Spotify is the best of a shambolic bunch. But I am pleased to see Mopidy-Tidal is active and I'm sure they are doing the best they can with what is available for Tidal. Maybe @tehkillerbee can comment better regarding the platform's stability.

**Comment by tehkillerbee** [comment]
@kingosticks Mopidy-Tidal in its current form works reasonably well and I use it myself on a daily basis together with Mopidy and a basic web client/android app. But integration with most frontends is limited at the moment.

Mopidy-tidal relies heavily on the work done by tidalapi. Whenever the Tidal plugin stops working, it is usually related to Tidal changing their API keys, requiring the tidalapi backend to be updated. Usually, it gets fixed relatively quickly, unless it is a bigger API change such as when Tidal changed their logon process to OAuth. Unfortunately, as you said, Tidal does not provide an open API, and it is a constant battle of the author of tidalapi and others to keep Tidal working. Since I am not currently a maintainer of [tidalapi](https://github.com/tamland/python-tidal), I cannot say if tidalapi will continue to be maintained. 

I think the only alternative is Volumio with the (paid) official Tidal Connect plugin. That's a pass for me.

I will try to play around with the Iris frontend and see how well it works together with Mopidy-tidal.

**Comment by tehkillerbee** [comment]
I got to play around a little with Mopidy-Iris this weekend. Main issue I experienced was related to issues with missing album art and albums that has been removed by Tidal, resulting in an error when browsing. I have tried to add some additional error handling to circumvent these issues. Feel free to test the latest Mopidy-Tidal 0.2.7.

- It takes quite some time to display the album art. Maybe this can be improved further with additional caching. Its especially visible when running on a RPi3.
- When browsing "Albums", it looks like album art is requested twice, once for individual URI's and then as a list of identical(?) URI's.
- Sometimes it is necessary to reload/refresh the folder, before album art is displayed.
- When browsing using the Iris "Playlists"/"Artists" shortcuts, album art is not shown. When browsing using "Tidal/My Playlists", "Tidal/My Artists", album art is displayed.

I'll probably continue testing Mopidy-Iris; it does look much better than the frontend I've been using up to this point!
